### DISCUSSION TITLE
Generic blocks in a class.

### DISCUSSION THREAD

**Discussion by cordasfilip** [dotnet/csharplang#5698] [Language Ideas]
Idea is to have a generic block so you can add fields associated to generic methods something similar to this.

```
class Table
{
     <T1>
         Dictionary<string, T1> field = new Dictionary<string, T1>();
         public T1 Get<T1>(string key)=>field[key];
         public void Set<T1>(string key, T1 value)=>field[key] = value;
     <T1>
}

var table = new Table();
table.Set("A", "TEST");
tableSet("B", 123);
table.Get<string>("A");
table.Get<int>("B");
``` 
This would avoid boxing for structs. There might be scenarios where this could be useful as well maybe buffers. Should work the same as it works when you put it on a class it should just apply to the same class.

**Comment by HaloFour** [comment]
How exactly would that work?  What type is `field`?  What happens if you do `table.Get<int>("A")`?

**Reply by cordasfilip** [reply]
> How exactly would that work? What type is `field`? What happens if you do `table.Get<int>("A")`?

How I see it if you invoke any of the methods in a block you generate all the methods even if they are not used.
The type are associated with the method invocations and all the methods must have the same Generic type parameters.

So the the 
`table.Get<int>("A")`
`table.Get<string>("A")`
would result in something similar to  
```
     class Table
     {
             Dictionary<string, int> field = new Dictionary<string, int>();
             public T1 Get<int>(string key)=>field[key];
             public void Set<int>(string key, int value)=>field[key] = value;

            Dictionary<string, string> field = new Dictionary<string, string>();
            public T1 Get<string>(string key)=>field[key];
            public void Set<string>(string key, string value)=>field[key] = value;
     }
```

**Reply by HaloFour** [reply]
> How I see it if you invoke any of the methods in a block you generate all the methods even if they are not used.

Generics don't work like this, they are not templates that are expanded at compile time.  Generics are expected to work across multiple assemblies and that's definitely not possible with this model.

**Reply by cordasfilip** [reply]
You are right generic might not be the best way to call it. The better way to say it would probably be generator templates not to be mixed with generics, I just would like to be able to do something similar. So I am probably asking for some way to do generic templates I just like the syntax. But it would probably be a lot of work for something that might not get used that often unless some way to create class templates was introduced .

**Comment by pinkfloydx33** [comment]
I don't see how you expect the dictionary to work. If the parameter is on the type that's one thing, but otherwise it seems to imply different/multiple instantiations of the dictionary.

**Reply by cordasfilip** [reply]
"it seems to imply different/multiple instantiations of the dictionary." Yes it would create a new field for each type

**Reply by pinkfloydx33** [reply]
How exactly do you propose that would work?

**Reply by cordasfilip** [reply]
Well generic would not be the right way to call it probably a better name would be template generated generic classes. My guess is how this could work is to have one additional assembly for all the generated classes for each assembly that used this classes in code(the original assembly). It would have to combine all the calls from all the assemblies that use it and combine it into this new class at build time. For the initial compile the methods would be throwing not implemented exception or something like that and when the class gets  seeded  new methods and fields are injected into the assembly.

**Reply by HaloFour** [reply]
It's not possible to know all of the code that will call it, even as a part of the current process.
